File: src/main/java/dev/mazurkiewicz/quizer/question/application/ApiQuestionService.java
package dev.mazurkiewicz.quizer.question.application;

import dev.mazurkiewicz.quizer.question.domain.model.*;
import dev.mazurkiewicz.quizer.question.domain.port.QuestionService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
class ApiQuestionService {

    private final QuestionService service;

    QuestionResponse getRandomQuestion() {
        final Question randomQuestion = service.getRandomQuestion();
        return QuestionResponse.of(randomQuestion);
    }

    QuestionResponse getQuestionById(Integer id) {
        QuestionId questionId = new QuestionId(id);
        Question question = service.getQuestion(questionId);
        return QuestionResponse.of(question);
    }

    public QuestionInfoResponse getQuestionsInfo() {
        QuestionInfo questionInfo = service.getQuestionInfo();
        return QuestionInfoResponse.of(questionInfo);
    }

    public AnswerStatusResponse checkAnswer(SelectedAnswerRequest selectedAnswerRequest) {
        QuestionId questionId = QuestionId.of(selectedAnswerRequest.questionId());
        SelectedAnswer selectedAnswer = SelectedAnswer.of(selectedAnswerRequest.selectedAnswer());
        AnswerResult answerResult = service.checkAnswer(questionId, selectedAnswer);
        return new AnswerStatusResponse(questionId.value(), answerResult.status(), answerResult.correctAnswer());
    }
}

File: src/main/kotlin/io/mazurkiewicz/quizer/question/application/ApiQuestionService.kt
package io.mazurkiewicz.quizer.question.application

import io.mazurkiewicz.quizer.question.domain.model.AnswerType
import io.mazurkiewicz.quizer.question.domain.model.QuestionId
import io.mazurkiewicz.quizer.question.domain.model.SelectedAnswer
import io.mazurkiewicz.quizer.question.domain.port.QuestionService
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class ApiQuestionService(private val questionService: QuestionService) {

    fun getRandomQuestion(): QuestionResponse {
        val randomQuestion = questionService.getRandomQuestion()
        return QuestionResponse(randomQuestion)
    }

    fun getQuestionById(id: UUID): QuestionResponse {
        val questionId = QuestionId(id)
        val question = questionService.getQuestion(questionId)
        return QuestionResponse(question)
    }

    fun checkAnswer(requestQuestionId: UUID, requestSelectedAnswer: String): AnswerStatusResponse {
        val questionId = QuestionId(requestQuestionId)
        val selectedAnswer = SelectedAnswer(AnswerType.of(requestSelectedAnswer))
        val answerResult = questionService.checkAnswer(questionId, selectedAnswer)
        return AnswerStatusResponse(questionId.value, answerResult.status, answerResult.correctAnswer)
    }
}